### PR TITLE
feat: add IntersectionJoiner

### LINKED ISSUE #55: Add BaseJoiner

### PR DESCRIPTION
**Pull request by piotrgramacki**
Closes #55

### CODE REVIEW

File: srai/joiners/intersection_joiner.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import pandas as pd
+
+
+@dataclass
+class JoinerResult:
+    """Joiner result dataclass."""
+
+    regions: gpd.GeoDataFrame
+    features: gpd.GeoDataFrame
+    joined: gpd.GeoDataFrame
```
Comment (RaczeQ) [inline srai/joiners/intersection_joiner.py:20]: This Joiner result could be extracted to a dedicated file since it will probably be shared between most joiners.

Comment (RaczeQ) [inline srai/joiners/intersection_joiner.py:20]: Maybe name this attribute `joint`? Or maybe make `joined` a more verbose name. It doesn't sound well IMO.

File: srai/joiners/intersection_joiner.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+              the intersection
+
+        """
+        joined_parts = []
+
+        for _, single in features.groupby(features["geometry"].geom_type):
+            joined_parts.append(
+                gpd.overlay(
+                    single[["geometry"]].reset_index().rename(columns={"index": "features_id"}),
+                    regions[["geometry"]].reset_index().rename(columns={"index": "hex_id"}),
```
Comment (RaczeQ) [inline srai/joiners/intersection_joiner.py:55]: `hex_id` should be renamed to `region_id`.

**Review by RaczeQ — CHANGES_REQUESTED** [review]

File: srai/joiners/intersection_joiner.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                    how="intersection",
+                    keep_geom_type=False,
+                ).set_index(["hex_id", "features_id"])
+            )
+
+        joined = gpd.GeoDataFrame(pd.concat(joined_parts, ignore_index=True))
```
[the remark this answers is not in the document]
Comment (piotrgramacki) [reply srai/joiners/intersection_joiner.py:61]: As long as input gdf's have unique indexes, there will be no overlap in MultiIndex in those concatenated gdf's. Am I right?

[the remark this answers is not in the document]
Comment (piotrgramacki) [reply srai/joiners/intersection_joiner.py:61]: But I guess there is no need to set `ignore_index=True`, so I'll change to default

File: srai/joiners/intersection_joiner.py (hunk 2 above)
Comment (piotrgramacki) [reply srai/joiners/intersection_joiner.py:55]: You're right, will fix it

File: srai/joiners/intersection_joiner.py (hunk 1 above)
Comment (piotrgramacki) [reply srai/joiners/intersection_joiner.py:20]: You're right, moving it to `base.py`

File: srai/joiners/intersection_joiner.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+              the intersection
+
+        """
+        joined_parts = []
+
+        for _, single in features.groupby(features["geometry"].geom_type):
+            joined_parts.append(
+                gpd.overlay(
+                    single[["geometry"]].reset_index().rename(columns={"index": "features_id"}),
+                    regions[["geometry"]].reset_index().rename(columns={"index": "region_id"}),
```
Comment (piotrgramacki) [inline srai/joiners/intersection_joiner.py:46]: Here I assume that both gdf's have unnamed index with correct values (as we decided to use index). Should I remove the name if provided (but this will change input gdf) or just assume that gdf is correct?

File: srai/joiners/intersection_joiner.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                    how="intersection",
+                    keep_geom_type=False,
+                ).set_index(["region_id", "features_id"])
+            )
+
+        joined = gpd.GeoDataFrame(pd.concat(joined_parts, ignore_index=False))
+
+        return JoinerResult(
```
Comment (simonusher) [inline srai/joiners/intersection_joiner.py:54]: Why do we want to return all 3 gdfs here?

File: srai/joiners/intersection_joiner.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            JoinerResult object, which contains 3 GeoDataFrames:
+            * original regions
+            * original features
+            * intersection of regions and features, which contains a MultiIndex and a geometry with
+              the intersection
+
+        """
+        joined_parts = []
+
+        for _, single in features.groupby(features["geometry"].geom_type):
```
Comment (simonusher) [inline srai/joiners/intersection_joiner.py:42]: What problem does this solve? Would a single spatial join not work?

Comment (piotrgramacki) [reply srai/joiners/intersection_joiner.py:42]: I could not force an `sjoin` into returning an intersection in a geometry field, but this may be my mistake. And `overlay` is limited to single geometry type in a whole gdf.

Comment (simonusher) [reply srai/joiners/intersection_joiner.py:42]: Oh, ok. I thought we wanted to only do sjoins here, and leave the potential processing (like intersecting geometries) to something further down the pipeline.

File: srai/joiners/intersection_joiner.py (hunk 5 above)
Comment (piotrgramacki) [reply srai/joiners/intersection_joiner.py:54]: If we process the whole pipeline probably it is not needed. But we decided to return all features here, and I followed @RaczeQ's suggestion to return it in this form (https://github.com/srai-lab/srai/issues/55#issuecomment-1296146540)

File: srai/joiners/intersection_joiner.py (hunk 6 above)
Comment (piotrgramacki) [reply srai/joiners/intersection_joiner.py:42]: Maybe I jump ahead with an intersection, then single `sjoin` will do the trick ;)

Comment (simonusher) [reply srai/joiners/intersection_joiner.py:42]: No, I think this is fine. I just needed an explanation

File: srai/joiners/intersection_joiner.py (hunk 4 above)
Comment (RaczeQ) [reply srai/joiners/intersection_joiner.py:46]: I'd make it as robust as possible. Best option would be to rename the index if it intervenes with this operation like this: `df.reset_index().rename(columns={df.index.name:'bar'})`.
https://stackoverflow.com/a/40914267/7766101

Comment (piotrgramacki) [reply srai/joiners/intersection_joiner.py:46]: The problem is, that when you apply `.reset_index()` when there is no name, the new column will be named `"index"`, otherwise `df.index.name`. And you cannot tell if something went wrong, because `.rename()` will not tell that it could not find a particular column.

An option is to remove name before, but this may either alter original gdf of increase memory consumption. I could also add second element to this mapping (`.rename(columns={"index": "bar", df.index.name: "bar"})`, but this may be problematic in some edge cases, where there is a column named "index"

Comment (RaczeQ) [reply srai/joiners/intersection_joiner.py:46]: You can rename it on the fly according to the documentation:

```python
single[["geometry"]].reset_index(names="features_id")
regions[["geometry"]].reset_index(names="region_id")
```

It will fail if there is already a column named like that or if the index is multileveled.

Comment (RaczeQ) [reply srai/joiners/intersection_joiner.py:46]: Additionally, `features` probably should be singularized into `feature_id`.

**Review by RaczeQ — CHANGES_REQUESTED** [review]
The code seems ok, but I'd add some tests (what happens with empty GDFs, empty on the left or right side, proper usage, disjointed geometries etc) and an example jupyter notebook.

File: srai/joiners/intersection_joiner.py (hunk 6 above)
Comment (Calychas) [reply srai/joiners/intersection_joiner.py:42]: > I could not force an `sjoin` into returning an intersection in a geometry field

hmm, that's interesting. That's what sjoin should do imo

**Comment by Calychas** [comment]
Overall great work. Let's sync on this PR

File: srai/joiners/intersection_joiner.py (hunk 6 above)
Comment (piotrgramacki) [reply srai/joiners/intersection_joiner.py:42]: I would too, but I always get an original geometry from either one of the gdfs

File: srai/joiners/base.py [hunk 7]
```diff
@@ -1,36 +1,24 @@
...
         Args:
             regions (gpd.GeoDataFrame): regions with which features are joined
             features (gpd.GeoDataFrame): features to be joined
         Returns:
-            JoinerResult object, which contains 3 GeoDataFrames:
-            * original regions
-            * original features
-            * joined features with regions in a MultiIndex with a geometry column
+            GeoDataFrame with an intersection of regions and features, which contains
+            a MultiIndex and a geometry with the intersection
```
Comment (RaczeQ) [inline srai/joiners/base.py:21]: Are we going to keep the type in Return as well in docstrings? That's how autodocformatter generates the output.
```suggestion
    Returns:
            gpd.GeoDataFrame: GeoDataFrame with an intersection of regions and features, which contains
            a MultiIndex and a geometry with the intersection.
```

**Review by Calychas — DISMISSED** [review]

**Review by RaczeQ — APPROVED** [review]
I think those tests are ok ✔️
We can maybe extract the fixtures outside since some of them are reused in both joiners and regionizers.
Additionally, integration tests should be prepared in the future (loader -> regionizer -> joiner -> embedder)

**Review by Calychas — DISMISSED** [review]
Great work! For tests I would use the already existing fixtures in regionizers but move them directory up

**Comment by Calychas** [comment]
> Additionally, integration tests should be prepared in the future (loader -> regionizer -> joiner -> embedder)
@RaczeQ 
Can be as well loader -> regionizer -> embedder -> joiner, which is the case for highway2vec

**Review by Calychas — APPROVED** [review]
